<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate enrichment scores from scRNA-seq data — irGSEA.score • irGSEA</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calculate enrichment scores from scRNA-seq data — irGSEA.score" />
<meta property="og:description" content="Input a Seurat object or scRNA-seq matrix, calculate the enrichment scores of
AUCell, UCell, singscore and ssgsea. Then, return a Seurat object including these
score matrix." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">irGSEA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/chuiqin/irGSEA/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate enrichment scores from scRNA-seq data</h1>
    <small class="dont-index">Source: <a href='https://github.com/chuiqin/irGSEA/blob/master/R/main.R'><code>R/main.R</code></a></small>
    <div class="hidden name"><code>irGSEA.score.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Input a Seurat object or scRNA-seq matrix, calculate the enrichment scores of
AUCell, UCell, singscore and ssgsea. Then, return a Seurat object including these
score matrix.</p>
    </div>

    <pre class="usage"><span class='fu'>irGSEA.score</span><span class='op'>(</span>
  object <span class='op'>=</span> <span class='cn'>NULL</span>,
  assay <span class='op'>=</span> <span class='cn'>NULL</span>,
  slot <span class='op'>=</span> <span class='st'>"data"</span>,
  min.cells <span class='op'>=</span> <span class='fl'>3</span>,
  min.feature <span class='op'>=</span> <span class='fl'>0</span>,
  seeds <span class='op'>=</span> <span class='fl'>123</span>,
  ncores <span class='op'>=</span> <span class='fl'>4</span>,
  custom <span class='op'>=</span> <span class='cn'>F</span>,
  geneset <span class='op'>=</span> <span class='cn'>NULL</span>,
  msigdb <span class='op'>=</span> <span class='cn'>T</span>,
  species <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>,
  category <span class='op'>=</span> <span class='st'>"H"</span>,
  subcategory <span class='op'>=</span> <span class='cn'>NULL</span>,
  geneid <span class='op'>=</span> <span class='st'>"symbol"</span>,
  method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"AUCell"</span>, <span class='st'>"UCell"</span>, <span class='st'>"singscore"</span>, <span class='st'>"ssgsea"</span><span class='op'>)</span>,
  aucell.MaxRank <span class='op'>=</span> <span class='cn'>NULL</span>,
  ucell.MaxRank <span class='op'>=</span> <span class='cn'>NULL</span>,
  kcdf <span class='op'>=</span> <span class='st'>"Gaussian"</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Seurat object or scRNA-seq matrix</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p>Name of assay to use, defaults to the active assay. The parameter
works when a seurat object is input.</p></td>
    </tr>
    <tr>
      <th>slot</th>
      <td><p>Default data. The parameter works if a seurat object is input.</p></td>
    </tr>
    <tr>
      <th>min.cells</th>
      <td><p>The minimum detected cells per gene, default 3. The parameter
worsk if a scRNA-seq matrix is input.</p></td>
    </tr>
    <tr>
      <th>min.feature</th>
      <td><p>The minimum genes per cell, default 0. The parameter works
if a scRNA-seq matrix is input.</p></td>
    </tr>
    <tr>
      <th>seeds</th>
      <td><p>Default 123</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Default 4</p></td>
    </tr>
    <tr>
      <th>custom</th>
      <td><p>Default False. Set it to true when input own genesets.</p></td>
    </tr>
    <tr>
      <th>geneset</th>
      <td><p>Default NULL. Input own genesets as a list. Each element in
the list is a gene set. You can also specify positive and negative genes by
adding a + or - sign in special gene set.</p></td>
    </tr>
    <tr>
      <th>msigdb</th>
      <td><p>Default True. You can acquire the collection gene sets from
msigdb database. It will be ignored when custom is set to true.</p></td>
    </tr>
    <tr>
      <th>species</th>
      <td><p>Default Homo sapiens. Use msigdbr::msigdbr_show_species() to view
all available species. The parameter works if msigdb is True.</p></td>
    </tr>
    <tr>
      <th>category</th>
      <td><p>Default H. You can acquire the Hallmarker gene sets. Use
msigdbr::msigdbr_collections to view all available collections gene sets. The
parameter works if msigdb is True.</p></td>
    </tr>
    <tr>
      <th>subcategory</th>
      <td><p>Default NULL. The parameter works if msigdb is True.</p></td>
    </tr>
    <tr>
      <th>geneid</th>
      <td><p>Default symbol. Other options are "entrez" and "ensembl". The
parameter works if msigdb is True.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A vector. Default all method.</p></td>
    </tr>
    <tr>
      <th>aucell.MaxRank</th>
      <td><p>The threshold to calculate the AUC. Default only the top 5<!-- % -->
of the expressed genes are used to checks whether the gene set are within
the top 5<!-- % when it set to NULL. You can inputc special number, such as 1500, -->
 to change the threshold. The parameter works if "AUCell" is selected in "method".</p></td>
    </tr>
    <tr>
      <th>ucell.MaxRank</th>
      <td><p>Maximum number of genes to rank per cell. Above this
rank, a given gene is considered as not expressed. Default 1500 when it set
to NULL. You can input special number, such as 2000, to increase the rank
range. The parameter works if "UCell" is selected in "method".</p></td>
    </tr>
    <tr>
      <th>kcdf</th>
      <td><p>Default Gaussian if input expression values are continuous in
logarithmic scale. Other option is "Poisson" if input expression values are
integer counts. The parameter works if "ssgsea" is selected in "method".</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a Seurat object including score matrix.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># load PBMC dataset by R package SeuratData</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://satijalab.org/seurat'>Seurat</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://www.satijalab.org/seurat'>SeuratData</a></span><span class='op'>)</span>
<span class='co'># download 3k PBMCs from 10X Genomics</span>
<span class='fu'><a href='https://rdrr.io/pkg/SeuratData/man/InstallData.html'>InstallData</a></span><span class='op'>(</span><span class='st'>"pbmc3k"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"pbmc3k.final"</span><span class='op'>)</span>
<span class='va'>pbmc3k.final</span> <span class='op'>&lt;-</span> <span class='fu'>SeuratObject</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/SeuratObject/man/UpdateSeuratObject.html'>UpdateSeuratObject</a></span><span class='op'>(</span><span class='va'>pbmc3k.final</span><span class='op'>)</span>

<span class='co'># Seurat object</span>
<span class='va'>pbmc3k.final</span> <span class='op'>&lt;-</span> <span class='fu'>irGSEA.score</span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>pbmc3k.final</span>, assay <span class='op'>=</span> <span class='st'>"RNA"</span>, slot <span class='op'>=</span> <span class='st'>"data"</span>,
msigdb <span class='op'>=</span> <span class='cn'>T</span>, species <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>, category <span class='op'>=</span> <span class='st'>"H"</span>, geneid <span class='op'>=</span> <span class='st'>"symbol"</span>,
method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"AUCell"</span>, <span class='st'>"UCell"</span>, <span class='st'>"singscore"</span>, <span class='st'>"ssgsea"</span><span class='op'>)</span>, kcdf <span class='op'>=</span> <span class='st'>'Gaussian'</span><span class='op'>)</span>

<span class='co'># data.fram, matrix, or dgmatrix</span>
<span class='va'>pbmc3k.final2</span> <span class='op'>&lt;-</span> <span class='fu'>irGSEA.score</span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>pbmc3k.final</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>counts</span>,
assay <span class='op'>=</span> <span class='st'>"RNA"</span>, slot <span class='op'>=</span> <span class='st'>"data"</span>, min.cells <span class='op'>=</span> <span class='fl'>3</span>, min.feature <span class='op'>=</span> <span class='fl'>0</span>,
method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"AUCell"</span>, <span class='st'>"UCell"</span>, <span class='st'>"singscore"</span>, <span class='st'>"ssgsea"</span><span class='op'>)</span>, kcdf <span class='op'>=</span> <span class='st'>'Poisson'</span><span class='op'>)</span>

<span class='co'># custom geneset</span>
<span class='va'>markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='va'>markers</span><span class='op'>$</span><span class='va'>Tcell_gd</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"TRDC+"</span>, <span class='st'>"TRGC1+"</span>, <span class='st'>"TRGC2+"</span>, <span class='st'>"TRDV1+"</span>,<span class='st'>"TRAC-"</span>,<span class='st'>"TRBC1-"</span>,<span class='st'>"TRBC2-"</span><span class='op'>)</span>
<span class='va'>markers</span><span class='op'>$</span><span class='va'>Tcell_NK</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"FGFBP2+"</span>, <span class='st'>"SPON2+"</span>, <span class='st'>"KLRF1+"</span>, <span class='st'>"FCGR3A+"</span>, <span class='st'>"CD3E-"</span>,<span class='st'>"CD3G-"</span><span class='op'>)</span>
<span class='va'>markers</span><span class='op'>$</span><span class='va'>Tcell_CD4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CD4"</span>,<span class='st'>"CD40LG"</span><span class='op'>)</span>
<span class='va'>markers</span><span class='op'>$</span><span class='va'>Tcell_CD8</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CD8A"</span>,<span class='st'>"CD8B"</span><span class='op'>)</span>
<span class='va'>markers</span><span class='op'>$</span><span class='va'>Tcell_Treg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"FOXP3"</span>,<span class='st'>"IL2RA"</span><span class='op'>)</span>
<span class='va'>pbmc3k.final3</span> <span class='op'>&lt;-</span> <span class='fu'>irGSEA.score</span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>pbmc3k.final</span>, assay <span class='op'>=</span> <span class='st'>"RNA"</span>, slot <span class='op'>=</span> <span class='st'>"data"</span>,
custom <span class='op'>=</span> <span class='cn'>T</span>, geneset <span class='op'>=</span> <span class='va'>markers</span>, method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"AUCell"</span>, <span class='st'>"UCell"</span>, <span class='st'>"singscore"</span>, <span class='st'>"ssgsea"</span><span class='op'>)</span>,
kcdf <span class='op'>=</span> <span class='st'>'Gaussian'</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Chuiqin Fan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


